src/store/options.js

```diff
@@ -1,37 +1,86 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
 import { store } from 'hybrids';
 
 const UPDATE_ACTION_NAME = 'updateOptions';
 
+export function isUpdateOptionsMessage(message) {
+  return message?.action === UPDATE_ACTION_NAME;
+}
+
+export const DNR_IDS = chrome.runtime
+  .getManifest()
+  .declarative_net_request.rule_resources.map((r) => r.id);
+
 const Options = {
+  dnrRules: DNR_IDS.reduce((all, rule) => ({ ...all, [rule]: true }), {}),
   trackerWheelDisabled: false,
   wtmSerpReport: true,
   [store.connect]: {
     async get() {
-      const storage = await chrome.storage.local.get(['options']);
-      return storage.options || {};
+      const { options = {} } = await chrome.storage.local.get(['options']);
+      return options;
     },
-    async set(_, options, keys) {
-      const prevOptions = await this.get();
-      const nextOptions = {
-        ...prevOptions,
-        ...Object.fromEntries(keys.map((key) => [key, options[key]])),
-      };
+    async set(_, options) {
+      await chrome.storage.local.set({ options });
 
-      chrome.storage.local.set({ options: nextOptions });
-      chrome.runtime.sendMessage({ action: UPDATE_ACTION_NAME });
+      // Send update message to another contexts (background page / panel / options)
```

Comment:
Unfortunately, I had to move that side effect back to the set action. The observe method is called when the model is cleared and fetched (because of the update in another context), which would trigger the loop if this side effect would be in the `observe`.